pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leaderelection
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gardener/etcd-backup-restore/pkg/errors"
+	"github.com/gardener/etcd-backup-restore/pkg/etcdutil"
+	"github.com/sirupsen/logrus"
+)
+
+var emptyStruct struct{}
+
+// NewLeaderElector returns LeaderElector configurations.
+func NewLeaderElector(logger *logrus.Entry, etcdConnectionConfig *etcdutil.EtcdConnectionConfig, leaderElectionConfig *ElectionConfig, callbacks *LeaderCallbacks) (*LeaderElector, error) {
+	return &LeaderElector{
+		logger:               logger.WithField("actor", "leader-elector"),
+		EtcdConnectionConfig: etcdConnectionConfig,
+		CurrentState:         DefaultCurrentState,
+		StopCh:               make(chan struct{}),
+		ProbeEtcdch:          make(chan struct{}),
+		PrevElectionResult:   ElectionLost,
+		Config:               leaderElectionConfig,
+		Callbacks:            callbacks,
+	}, nil
+}
+
+// StartElection starts the LeaderElection to Elect the backup-restore's Leader.
+func (le *LeaderElector) StartElection(ctx context.Context) {
```

Comment:
```suggestion
// Run starts the LeaderElection to Elect the backup-restore's Leader.
func (le *LeaderElector) Run(ctx context.Context) {
```

`StartElection` means one election. But here it is a continuous series of elections. So `Run` is a better name for the function.